### PR TITLE
Experimental typeImports

### PR DESCRIPTION
**Pull request by drslump**
Changes the behaviour of `from m import T` following the discussion at https://github.com/nim-lang/Nim/issues/8013

To be clear, since it's not exactly the same as discussed on the other issue, the behaviour implemented is:

 - Automatically include any symbol that references the imported type on its signature

That doesn't include generic types, basically because they are a tad more difficult to implement and I'm not knowledgeable enough to assert if it's something that should be allowed or not regardless of me being able to do the implementation. Comments about this more than welcome :)

### CODE REVIEW

File: compiler/importer.nim [hunk 1]
```diff
@@ -68,6 +68,22 @@ proc rawImportSymbol(c: PContext, s: PSym) =
...
+proc maybeImportForType(c: PContext, s: PSym, t: PType) =
+  ## Imports a symbol only if it has a strict relation with a type:
+  ##
+  ##  - Proc refers to the type on any of its input or output values.
+  ##
+  ## TODO: generics are not matched currently.
+  ##
+  case s.kind
+  of skProcKinds:
+    for tt in s.typ.sons.items:
```
Comment (drslump) [inline compiler/importer.nim]: Question: shall I avoid using iterators in the compiler code? maybe they are not as efficient as `countup` loops?

Comment (Araq) [reply compiler/importer.nim]: No need to avoid them, they are zero-cost.

File: doc/manual.rst [hunk 2]
```diff
@@ -6333,6 +6333,24 @@ It's also possible to use ``from module import nil`` if one wants to import
...
+
+.. code-block:: nim
+    :test: "nim c $1"
+
+  {.experimental: "typeImports".}
+
+  from colors import Color
+
+  # to string and proc imported
+  let pink = rgb(255, 192, 203)
```
Comment (Araq) [inline doc/manual.rst]: I think this example shows that it's not enough the type is the return type of ``rgb``.

File: compiler/importer.nim [hunk 3]
```diff
@@ -68,6 +68,22 @@ proc rawImportSymbol(c: PContext, s: PSym) =
     if s.kind == skConverter: addConverter(c, s)
     if hasPattern(s): addPattern(c, s)
 
+proc maybeImportForType(c: PContext, s: PSym, t: PType) =
+  ## Imports a symbol only if it has a strict relation with a type:
+  ##
+  ##  - Proc refers to the type on any of its input or output values.
+  ##
+  ## TODO: generics are not matched currently.
```
Comment (Araq) [inline compiler/importer.nim]: Yeah, this should be done though. Requires a bit more code, will show you how to do that later.

File: doc/manual.rst (hunk 2 above)
Comment (drslump) [reply doc/manual.rst]: sorry, I don't understand, could you rephrase?

Maybe it should demonstrate what's going on more explicitly `let pink: Color = rgb(255, 192, 203)`?

Comment (Araq) [reply doc/manual.rst]: I mean that ``rgb`` should not be imported implicitly as the connection to ``Color`` is invisible.

**Comment by timotheecour** [comment]
IMO the feature should not work with regular import syntax:
`from colors import Color`
as a user could genuinely not want a particular method to be imported (even with UFCS)

instead, should be explicit (that way, doens't even have to be "experimental":
`from colors import Color.*` or, more explicit:
`import colors: withMethodsmethods(Color)`

**Comment by Araq** [comment]
This feature exists to bring Nim closer to Python and I don't think we should introduce more syntax. And even if we did, the ``.experimental`` would remain, new features need to start in the experimental state, no matter how good they seem to be.

**Comment by drslump** [comment]
Implemented support for generics by using `sigmatch.typeRel`, it works but I couldn't say if there is any performance implications about the solution, I'm way in over my head to be honest :-)

**Comment by Araq** [comment]
1. Use the ``typeRel`` mechanism for the first parameter only, not for return types or anything else.
2. Yes this means ``from tables import Table, initTable`` but this is the better design. The feature exists because you argue it's not obvious where stuff "comes from" and it isn't obvious for ``rgb``.

Otherwise every proc in e.g. ``regex`` deals with the ``Regex`` type and so ``import re`` is the same as ``from re import Regex``. And arguably that's already the case for most modules out there, making this whole feature solving a non-issue, but hey, we already know that.

**Comment by drslump** [comment]
gotcha, so long, and thanks for all the fish
